### PR TITLE
feat: provide formatting subcommand 

### LINKED ISSUE #1580: Nargo: Provide formatting subcommand
### Problem

Nargo should provide a `nargo format` command that will format files and write them back to a filesystem.

### Happy Case

A user can write code in whatever format they want, but running `nargo format` will always produce the same output.

### Alternatives Considered

_No response_

### Additional Context

See also:
- prettier
- cargo fmt
- grain format

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by f01dab1e**
# Description

WIP

## Problem 

Resolves #1580

## Summary

WIP

# PR Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_frontend/src/parser/mod.rs [hunk 1]
```diff
@@ -217,6 +217,85 @@ fn force<'a, T: 'a>(parser: impl NoirParser<T> + 'a) -> impl NoirParser<Option<T
     parser.map(Some).recover_via(empty().map(|_| None))
 }
 
+pub trait Mode: std::fmt::Debug + Clone + Copy + 'static {
```
Comment (kevaundray) [inline crates/noirc_frontend/src/parser/mod.rs:220]: why is this needed?

**Review by kevaundray — COMMENTED** [review]
The fact that we are parsing in "compiler mode" has leaked into the later passes, so I'd like to question the reason for this and how/if cargo fmt, prettier etc do this

File: crates/noirc_frontend/src/parser/mod.rs (hunk 1 above)
Comment (f01dab1e) [reply crates/noirc_frontend/src/parser/mod.rs:220]: I added it because I need to preserve spans and the order of items. Perhaps it can be removed and just slightly modify the AST, but for now, I don't want to change the compiler.

**Comment by f01dab1e** [comment]
rustfmt: https://github.com/rust-lang/rustfmt/blob/master/Design.md#implementation-philosophy
CST-based formatter: https://github.com/rust-lang/rust-analyzer/issues/1665

File: crates/noirc_frontend/src/parser/mod.rs (hunk 1 above)
Comment (kevaundray) [reply crates/noirc_frontend/src/parser/mod.rs:220]: When you say for now, do you mean while the PR is in draft mode? If so, then I think thats okay while you flesh out the rest of the formatting code

Comment (f01dab1e) [reply crates/noirc_frontend/src/parser/mod.rs]: removed

File: tooling/nargo_fmt/tests/expected/comment.nr [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// random comment
+
+fn comment2() {
+// Test
+}
+fn comment3() {}
+fn comment4() {}
+fn comment5() {}
+fn comment6() {}
+fn comment7() {}
```
Comment (jfecher) [inline tooling/nargo_fmt/tests/expected/comment.nr:14]: A `fmt` command as a whole is greatly appreciated, but I'm going to block this PR while comments are being deleted. I think a `fmt` command that deletes comments would be a deal breaker for most users and don't want them using the command and accidentally deleting comments in the meantime while we work on it.

Preserving comments can be difficult in general though so we'd need some work in adding them to our syntax tree.

File: tooling/nargo_fmt/tests/expected/expr.nr [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            }
+        }
+    }
+}
+fn comment() {
+    // this is a test comment
+        1
+}
+fn only_comment() {
+// Keep this here
```
Comment (jfecher) [inline tooling/nargo_fmt/tests/expected/expr.nr:34]: It looks like we expect an extra indent on the 1 and no extra newline between these two functions. Are the "input" and "expected" files switched for this test?

File: tooling/nargo_fmt/tests/expected/nested-if-else.nr [hunk 4]
```diff
@@ -0,0 +1,11 @@
+fn nested_if_else() {
+    if false {
+    1
+} else {
```
Comment (jfecher) [inline tooling/nargo_fmt/tests/expected/nested-if-else.nr:4]: Some indentation to improve upon here as well. Since this PR is still a draft, I'm assuming this is already known and I'll stop pointing it out.

**Review by jfecher — CHANGES_REQUESTED** [review]
Blocking pending preserving all comments in source code and some indentation changes in the formatter.

I wish I had a cleaner suggestion for preserving comments in a formatter, but the main solution I'm aware of is to add comment nodes in a concrete syntax tree. This tree would have less structure than our AST in that any node would take a `Vec<Node>` as children, which includes comments and any actual children (e.g. each statement for a block).

**Comment by kevaundray** [comment]
@f01dab1e Is this ready for re-review?

**Comment by f01dab1e** [comment]
@kevaundray I think yes

File: compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs [hunk 5]
```diff
@@ -461,8 +461,9 @@ impl<'a> ModCollector<'a> {
         context.visited_files.insert(child_file_id, location);
 
         // Parse the AST for the module we just found and then recursively look for it's defs
-        //let ast = parse_file(&context.file_manager, child_file_id, errors);
+        //let ast = parse_file(&context.file_manager, child_file_id, errors).into_unorder();
```
Comment (kevaundray) [inline compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs:464]:
```suggestion
```

**Comment by kevaundray** [comment]
How does this preserve comments?

**Comment by f01dab1e** [comment]
@kevaundray 

we save the last position of the node fragment, then take the slice...

the `format_missing_inner` method does this.

for example:
```rust
// test
fn main() {}
```

Call the function with the starting position of the function (8)
1. self.format_missing_indent(span.start(), false);
1.2 take the slice and write it to the buffer (source[self.last_position, end] // end = span.start()
1.3 write 


for the rest it happens in push_rewrite
```
#[track_caller]
fn push_rewrite(&mut self, s: String, span: Span) {
    self.format_missing_indent(span.start(), true);
    self.push_str(&s);
}
```

Issues arise as for some vertices we don't have a span and will need to conduct a repeated lexical analysis to search for the required symbol in order to take a slice with the comment.

File: compiler/noirc_frontend/src/parser/mod.rs [hunk 6]
```diff
@@ -236,6 +235,55 @@ pub struct ParsedModule {
...
 }
 
+/// A ParsedModule contains an entire Ast for one file.
+#[derive(Clone, Debug, Default)]
+pub struct ParsedModule {
+    pub items: Vec<Item>,
+}
+
+impl ParsedModule {
+    pub fn into_unorder(self) -> UnorderParsedModule {
```
Comment (jfecher) [inline compiler/noirc_frontend/src/parser/mod.rs:245]: Ok, now I see it means it is the separated kind. Perhaps rename Unorder[ed] to `SortedModule` then.

**Review by jfecher — DISMISSED** [review]
Code looks good. I only skimmed the new additions in the new fmt crate but the changes to the Ast and ParsedModule seem good to me, and it seems the formatter can handle comments as well now. I saw a somewhat questionable macro in the fmt code but it seemed somewhat helpful so it can stay.

File: tooling/nargo_fmt/src/visitor/expr.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+                visitor.block_indent = self.block_indent;
+                visitor.visit_block(block, span, true);
+
+                visitor.buffer
+            }
+            ExpressionKind::Prefix(prefix) => {
+                format!("{}{}", prefix.operator, self.format_expr(prefix.rhs))
+            }
+            // TODO:
```
Comment (kevaundray) [inline tooling/nargo_fmt/src/visitor/expr.rs:28]: create an issue to fix

**Review by kevaundray — APPROVED** [review]
I think we can iterate towards something that works for both the AST mode and the CST mode -- can you document in nargo format how comments are being handled?

File: tooling/nargo_cli/src/cli/mod.rs [hunk 8]
```diff
@@ -52,6 +53,7 @@ pub(crate) struct NargoConfig {
 enum NargoCommand {
     Backend(backend_cmd::BackendCommand),
     Check(check_cmd::CheckCommand),
+    Fmt(fmt_cmd::FormatCommand),
```
Comment (kevaundray) [inline tooling/nargo_cli/src/cli/mod.rs:56]:
```suggestion
    #[command(hide = true)] // Hidden while the feature has not been extensively tested
    Fmt(fmt_cmd::FormatCommand),
```

Comment (kevaundray) [reply tooling/nargo_cli/src/cli/mod.rs:56]: Lets make the feature there but keep it hidden while we test it out a bit and make modifications

**Comment by kevaundray** [comment]
For a future PR, it seems that Rust does not store the comment as a node but instead attaches comments to Statements and Expressions, so something along the lines of every expression/statement having an optional comment field.

This is not entirely how it works since Rust has a few other abstractions but this gives us the gist

File: tooling/nargo_fmt/src/visitor/expr.rs (hunk 7 above)
Comment (f01dab1e) [reply tooling/nargo_fmt/src/visitor/expr.rs:28]: #2993

**Review by kevaundray — APPROVED** [review]
